grakn-graph/src/main/java/ai/grakn/graph/internal/RoleTypeImpl.java

```diff
@@ -77,10 +100,17 @@
      */
     @Override
     public Collection<Type> playedByTypes() {
-        Set<Type> playedBy = new HashSet<>();
-        getIncomingNeighbours(Schema.EdgeLabel.PLAYS_ROLE).
-                forEach(concept -> playedBy.addAll(concept.asType().subTypes()));
-        return playedBy;
+        Set<Type> playedByTypes = new HashSet<>();
+        cachedDirectPlayedByTypes.get().forEach(type -> playedByTypes.addAll(type.subTypes()));
+        return Collections.unmodifiableCollection(playedByTypes);
+    }
+
+    void addCachedDirectPlaysByType(Type newType){
+        cachedDirectPlayedByTypes.get().add(newType);
```

Comment:
Possibly these methods should only be adding to the cache if it already has been cached? Otherwise they can sneakily do nothing.